rats-tls/src/core/cpu.c

```diff
@@ -15,53 +15,109 @@
 static inline void cpuid(int *eax, int *ebx, int *ecx, int *edx)
 {
 #if defined(__x86_64__)
-    asm volatile("cpuid"
-                 : "=a"(*eax), "=b"(*ebx), "=c"(*ecx), "=d"(*edx)
-                 : "0"(*eax), "1"(*ebx), "2"(*ecx), "3"(*edx)
-                 : "memory");
+	asm volatile("cpuid"
+		     : "=a"(*eax), "=b"(*ebx), "=c"(*ecx), "=d"(*edx)
+		     : "0"(*eax), "1"(*ebx), "2"(*ecx), "3"(*edx)
+		     : "memory");
 #else
-    /* on 32bit, ebx can NOT be used as PIC code */
-    asm volatile("xchgl %%ebx, %1; cpuid; xchgl %%ebx, %1"
-                 : "=a"(*eax), "=r"(*ebx), "=c"(*ecx), "=d"(*edx)
-                 : "0"(*eax), "1"(*ebx), "2"(*ecx), "3"(*edx)
-                 : "memory");
+	/* on 32bit, ebx can NOT be used as PIC code */
+	asm volatile("xchgl %%ebx, %1; cpuid; xchgl %%ebx, %1"
+		     : "=a"(*eax), "=r"(*ebx), "=c"(*ecx), "=d"(*edx)
+		     : "0"(*eax), "1"(*ebx), "2"(*ecx), "3"(*edx)
+		     : "memory");
 #endif
 }
 
 static inline void __cpuidex(int a[4], int b, int c)
 {
-    a[0] = b;
-    a[2] = c;
-    cpuid(&a[0], &a[1], &a[2], &a[3]);
+	a[0] = b;
+	a[2] = c;
+	cpuid(&a[0], &a[1], &a[2], &a[3]);
+}
+
+static bool is_sgx_device(const char *dev)
+{
+	struct stat st;
+
+	if (!stat(dev, &st)) {
+		if ((st.st_mode & S_IFCHR) && (major(st.st_rdev) == 10))
+			return true;
+	}
+
+	return false;
+}
+
+static bool is_legacy_oot_kernel_driver(void)
+{
+	return is_sgx_device("/dev/isgx");
+}
+
+static bool is_dcap_oot_kernel_driver(void)
+{
+	return is_sgx_device("/dev/sgx/enclave");
```

Comment:
To my understanding, the Dcap's oot driver is the same as in-kernel driver on sgx node name in the new version.

Reply:
Exactly. But this happens since dcap oot driver 1.10. Prior to it, we'd better to have a forward-compatibility with the old dcap oot driver. I will add a comment and rename the function name to is_dcap19_oot_kernel_driver().